tests/components/homekit_controller/test_init.py

```diff
@@ -218,3 +227,57 @@ async def get_characteristics(self, chars, *args, **kwargs):
     is_available = True
     async_fire_time_changed(hass, utcnow() + timedelta(hours=1))
     assert hass.states.get("light.testdevice").state == STATE_OFF
+
+
+@pytest.mark.parametrize("example", FIXTURES, ids=lambda val: str(val.stem))
+async def test_snapshots(
+    hass: HomeAssistant, snapshot: SnapshotAssertion, example: str
+) -> None:
+    """Detect regressions in enumerating a homekit accessory database and building entities."""
+    accessories = await setup_accessories_from_file(hass, example)
+    config_entry, _ = await setup_test_accessories(hass, accessories)
+
+    device_registry = dr.async_get(hass)
+    entity_registry = er.async_get(hass)
+
+    registry_devices = dr.async_entries_for_config_entry(
+        device_registry, config_entry.entry_id
+    )
+    registry_devices.sort(key=lambda device: device.name)
+
+    devices = []
+
+    for device in registry_devices:
+        entities = []
+
+        registry_entities = er.async_entries_for_device(
+            entity_registry,
+            device_id=device.id,
+            include_disabled_entities=True,
+        )
+        registry_entities.sort(key=lambda entity: entity.entity_id)
+
+        for entity_entry in registry_entities:
+            state_dict = None
+            if state := hass.states.get(entity_entry.entity_id):
+                state_dict = dict(state.as_dict())
+                state_dict.pop("context", None)
+                state_dict.pop("last_changed", None)
+                state_dict.pop("last_updated", None)
+
+                state_dict["attributes"] = dict(state_dict["attributes"])
+                state_dict["attributes"].pop("access_token", None)
+                state_dict["attributes"].pop("entity_picture", None)
+
+            entry = asdict(entity_entry)
+            entry.pop("id", None)
+            entry.pop("device_id", None)
+
+            entities.append({"entry": entry, "state": state_dict})
+
+        device_dict = asdict(device)
+        device_dict.pop("id", None)
+        device_dict.pop("via_device_id", None)
+        devices.append({"device": device_dict, "entities": entities})
+
+    assert snapshot == devices
```

Comment:
fyi, you can also ignore stuff on syrupy level, not sure if that helps, but maybe :)

Reply:
I saw thanks. I wasn't sure it made things better than this way, and I'm not worried about code reuse here, so didn't seem worth it. If there's a particular example that you think would make this more readable or cleaner I'd love to see it though.

Reply:
I think I once used it to ignore entry_id of config entries. But I think since you're double looping, this solution probably looks best.